crontab.1
.\" Automatically generated by Pandoc 2.3.1
.\"
.TH "CRONTAB" "1" "20 Nov 2019" "" ""
.hy
.SH NAME
.PP
crontab \- manipulate per\-user crontabs (dillon's lightweight cron
daemon)
.SH SYNOPSIS
.PP
\f[B]crontab file [\-u user]\f[] \- replace crontab from file
.PP
\f[B]crontab \- [\-u user]\f[] \- replace crontab from stdin
.PP
\f[B]crontab \-l [\-u user]\f[] \- list crontab for user
.PP
\f[B]crontab \-e [\-u user]\f[] \- edit crontab for user
.PP
\f[B]crontab \-d [\-u user]\f[] \- delete crontab for user
.PP
\f[B]crontab \-c dir\f[] \- specify crontab directory
.SH DESCRIPTION
.PP
The \f[B]crontab\f[] command manipulates the per\-user crontabs.
If you are looking for the format of these files, see \[lq]man 5
crontab\[rq].
.PP
Generally the \-e option is used to edit your crontab.
\f[B]crontab\f[] will use the editor specified by your EDITOR or VISUAL
environment variable (or /usr/bin/vi) to edit the crontab:
.IP
.nf
\f[C]
$\ crontab\ \-e
\f[]
.fi
.PP
Just as a regular user, the superuser also has his or her own per\-user
crontab.
.SH NOTES
.PP
\f[B]crontab\f[] doesn't provide the kinds of protections that programs
like \f[B]visudo\f[] do against syntax errors and simultaneous edits.
Errors won't be detected until \f[B]crond\f[] reads the crontab file.
What \f[B]crontab\f[] does is provide a mechanism for users who may not
themselves have write privileges to the crontab folder to nonetheless
install or edit their crontabs.
It also notifies a running crond daemon of any changes to these files.
.PP
Only users who belong to the same group as the \f[B]crontab\f[] binary
will be able to install or edit crontabs.
However it'll be possible for the superuser to install crontabs even for
users who don't have the privileges to install them themselves.
(Even for users who don't have a login shell.) Only the superuser may
use the \-u or \-c switches to specify a different user and/or crontab
directory.
.SH SEE ALSO
.PP
\f[B]crontab\f[](5) \f[B]crond\f[](8)
.SH AUTHORS
.PP
Matthew Dillon (dillon\@apollo.backplane.com): original developer
.PD 0
.P
.PD
James Pryor (dubiousjim\@gmail.com): current developer


CHANGELOG
todo
  * add timestamps to -d output

  * manual cron.update prodding doesn't affect cl_NotUntil

  * Use hash when ID=... is supplied. FS#18292.

  * FS#18352: Another thing: when moving the original file to the backup name, and the edited version is written in it's place, the file is written without preserving the same permissions as the original, so if you have a umask that prevents others from reading your stuff, crontab won't be able to load the new file.

git
  * Numeric loglevels specified by 'crond -l <level>' weren't being validated.
    Now we no longer accept numeric loglevels; they must be specified
    symbolically. Thanks to Rogutės Sparnuotos.

  * Continued portability improvements. Makefile now uses -lbsd-compat.
    Factored allocation and string calls to utils.c.

  * Added extra/crond.service for systemd. Thanks to Miklos Vajna.

  * Many internal changes and annotations to pass splint review.

  * Documentation and error message updates.

v4.5  1-May-2011
  * Some cron jobs were running multiple times. Now we make sure not to
    ArmJobs that are already running; and not to resynchronize while jobs are
    running; and to poll the DST setting. (Fixes Arch FS#18681; thanks to Vincent
    Cappe and Paul Gideon Dann for identifying the second issue; and Tilman
    Sauerbeck for identifying the third.)

  * @monthly was wrongly being parsed the same as @yearly (fixes Arch
    FS#19123). Thanks to Peter Johnson, Paul Gideon Dann, and Tilman Sauerbeck.

  * extra/crond.rc: now uses $CROND_ARGS from /etc/conf.d/crond; sample included
    as extra/crond.conf. Suggested by Eric Bélanger.

  * Running `/etc/rc.d/crond start` after startup could leak unwanted
    environment into cronjobs; now we force crond to start in empty env
    (fixes Arch FS#22085). Thanks to Mantas.

  * Also set LOGNAME environment variable in cronjobs. Requested by Michael
    Trunner; fixes Arch FS#18338.

  * extra/crond.logrotate now correctly gets pid from /var/run/crond.pid
    (fixes Arch FS#18039). Thanks to Kay Abendroth, revel, and Chlump Chatkupt.

  * extra/prune-cronstamps now only deletes files, and is formatted as a
    @weekly crontab. Thanks to Alec Moskvin <alecm@gmx.com>.

  * extra/crontab.vim works around an issue where vim's writebackup would
    interfere with crontab's security model (addresses Arch FS#18352).
    Thanks to Armadillo and Simon Williams.

  * Makefile uses $LDFLAGS (fixes Arch FS#23784). Thanks to Kristoffer Tidemann
    and Mike Frysinger.

  * defs.h sets default locations for CRONTABS and CRONSTAMPS beneath /var/spool/cron/,
    as in earlier versions of dcron.

  * Documentation updates.

  * Thanks for testing and feedback: Feifei Jia, Spider.007, Ray Kohler,
    Igor Zakharoff, Edward Hades, and Joe Lightning.

v4.4  17-Jan-2010
  * Finished mailjobs were being left as zombie processes. Fixed.

  * When using crond with logging-to-file, user jobs could only log some
    events if they had write access to the log. Fixed this by having crond
    keep a file descriptor open to the log; also added a SIGHUP handler
    to make crond re-open the logfile. The sample logrotate script now
    sends that signal.

  * More sensible command-line parsing by crontab.

  * Add prune-cronstamps to extra; document extra/*; general improvement
    of README and manpages.

  * Portability improvements, and defs.h now has fuller comments about
    requirements.

  * Makefile improvements: `make` now caches variables for `make install`;
    don't stomp CFLAGS environment variable, and added BINDIR,SBINDIR,MANDIR.

  * Thanks to Juergen Daubert for more testing and suggestions.

v4.3  11-Jan-2010
  * Internal refactoring to make buffer overflow checks
    clearer and portability issues more explicit.

  * Made file argument to -L mandatory; optional args to
    getopt needs GNU extensions.

  * Makefile tweaks. Added CRONTAB_GROUP for `make install`.
    Renamed TIMESTAMPS -> CRONSTAMPS.

  * Thanks to Juergen Daubert for testing and suggestions.

v4.2  11-Jan-2010
  * Makefile tweaks; moved more constants to #defines.

v4.1  10-Jan-2010
  * Fixed bug in parsing some numeric fields in crontabs. (Terminus of range
    wasn't being modded.)

  * Updated Makefile to make it easier to customize timestamps at configure
    time. Also, if LC_TIME is defined when crond runs, we use that instead of
    compiled-in default (for logging to files, to customize syslog output use
    syslog-ng's 'template' command).

  * Fixed Makefile permissions on crond and crontab binaries.

v4.0  6-Jan-2010
  * Jim Pryor took over development; folded in changes from his fork "yacron"

  * Applied "Daniel's patch" from dcron 3.x tarballs to enable logging to syslog or
    files. Added further logging improvements.

  * Added -m user@host and -M mailer options

  * Various crontab syntax extensions, including "2nd Monday of every month",
    @reboot, @daily, and finer-grained frequency specifiers.

  * Jobs can wait until AFTER other jobs have finished.

  * Enhanced parsing of cron.update file, to make it possible for scripts to
    interact with a running crond in limited ways.

  * Various internal changes

  * Updated Makefile, manpage buildchain, and docs

v3.2
    Fixed a minor bug, remove the newline terminating a line only if there
    is a newline to remove.

v3.1
    Add support for root-run crontab files in /etc/cron.d and rewrite a
    good chunk of the crontab file management code.  By VMiklos and Matt
    Dillon.

v3.0
    Fix /tmp race and misc cleanups from Emiel Kollof <emiel@gamepoint.net>

v2.9
    Modernize the code, remove strcpy() and sprintf() in favor of snprintf().
    (Supplied by Christine Jamison <technobabe@mail.nwmagic.net>)

v2.8
    Fixed bug found by Christian HOFFMANN.  newline removal was broken
    for lines that began with whitespace, causing crontab lines to be
    chopped off.

v2.7
    Committed changes suggested by
    Ragnar Hojland Espinosa <ragnar@redestb.es>


crontab.c

/*
 * CRONTAB.C
 *
 * crontab [-u user] [-c dir] [-l|-e|-d|file|-]
 * usually run as setuid root
 * -u and -c options only work if getuid() == geteuid()
 *
 * Copyright 1994 Matthew Dillon (dillon@apollo.backplane.com)
 * Copyright 2009-2019 James Pryor <dubiousjim@gmail.com>
 * May be distributed under the GNU General Public License version 2 or any later version.
 */

#include "defs.h"

Prototype void printlogf(int level, const char *ctl, ...);

void Usage(void);
int GetReplaceStream(const char *user, const char *file);
void EditFile(const char *user, const char *file);

const char *CDir = CRONTABS;
int   UserId;


int
main(int ac, char **av)
{
	enum { NONE, EDIT, LIST, REPLACE, DELETE } option = NONE;
	struct passwd *pas;
	char *repFile = NULL;
	int repFd = 0;
	int i;
	char caller[SMALL_BUFFER];		/* user that ran program */

	UserId = getuid();
	if ((pas = getpwuid(UserId)) == NULL) {
		perror("getpwuid");
		exit(1);
	}
	/* [v]snprintf write at most size including \0; they'll null-terminate, even when they truncate */
	/* return value >= size means result was truncated */
	if (snprintf(caller, sizeof(caller), "%s", pas->pw_name) >= sizeof(caller)) {
		printlogf(0, "username '%s' too long", caller);
		exit(1);
	}

	opterr = 0;
	while ((i=getopt(ac,av,"ledu:c:")) != -1) {
		switch(i) {
			case 'l':
				if (option != NONE)
					Usage();
				else
					option = LIST;
				break;
			case 'e':
				if (option != NONE)
					Usage();
				else
					option = EDIT;
				break;
			case 'd':
				if (option != NONE)
					Usage();
				else
					option = DELETE;
				break;
			case 'u':
				/* getopt guarantees optarg != 0 here */
				if (*optarg != 0 && getuid() == geteuid()) {
					pas = getpwnam(optarg);
					if (pas) {
						UserId = pas->pw_uid;
						/* paranoia */
						if ((pas = getpwuid(UserId)) == NULL) {
							perror("getpwuid");
							exit(1);
						}
					} else {
						printlogf(0, "user '%s' unknown", optarg);
						exit(1);
					}
				} else {
					printlogf(0, "-u option: superuser only");
					exit(1);
				}
				break;
			case 'c':
				/* getopt guarantees optarg != 0 here */
				if (*optarg != 0 && getuid() == geteuid()) {
					CDir = optarg;
				} else {
					printlogf(0, "-c option: superuser only");
					exit(1);
				}
				break;
			default:
				/* unrecognized -X */
				option = NONE;
		}
	}

	if (option == NONE && optind == ac - 1) {
		if (av[optind][0] != '-') {
			option = REPLACE;
			repFile = av[optind];
			optind++;
		} else if (av[optind][1] == 0) {
			option = REPLACE;
			optind++;
		}
	}
	if (option == NONE || optind != ac) {
		Usage();
	}

	/*
	 * If there is a replacement file, obtain a secure descriptor to it.
	 */

	if (repFile) {
		repFd = GetReplaceStream(caller, repFile);
		if (repFd < 0) {
			printlogf(0, "unable to read replacement file %s", repFile);
			exit(1);
		}
	}

	/*
	 * Change directory to our crontab directory
	 */

	if (chdir(CDir) < 0) {
		printlogf(0, "cannot change dir to %s: %s", CDir, strerror(errno));
		exit(1);
	}

	/*
	 * Handle options as appropriate
	 */

	switch(option) {
		case LIST:
			{
				FILE *fi;
				char buf[RW_BUFFER];

				if ((fi = fopen(pas->pw_name, "r"))) {
					while (fgets(buf, sizeof(buf), fi) != NULL)
						fputs(buf, stdout);
					fclose(fi);
				} else {
					fprintf(stderr, "no crontab for %s\n", pas->pw_name);
					/* no error code */
				}
			}
			break;
		case EDIT:
			{
				FILE *fi;
				int fd;
				int n;
				char tmp[] = TMPDIR "/crontab.XXXXXX";
				char buf[RW_BUFFER];

				/*
				 * Create temp file with perm 0600 and O_EXCL flag, ensuring that this call creates the file
				 * Read from fi for "$CDir/$USER", write to fd for temp file
				 * EditFile changes user if necessary, and runs editor on temp file
				 * Then we delete the temp file, keeping its fd as repFd
				 */
				if ((fd = mkstemp(tmp)) >= 0) {
					chown(tmp, getuid(), getgid());
					if ((fi = fopen(pas->pw_name, "r"))) {
						while ((n = fread(buf, 1, sizeof(buf), fi)) > 0)
							write(fd, buf, n);
					}
					EditFile(caller, tmp);
					remove(tmp);
					lseek(fd, 0L, 0);
					repFd = fd;
				} else {
					printlogf(0, "unable to create %s: %s", tmp, strerror(errno));
					exit(1);
				}

			}
			option = REPLACE;
			/* fall through */
		case REPLACE:
			{
				char buf[RW_BUFFER];
				char path[SMALL_BUFFER];
				int fd;
				int n;

				/*
				 * Read from repFd, write to fd for "$CDir/$USER.new"
				 */
				snprintf(path, sizeof(path), "%s.new", pas->pw_name);
				if ((fd = open(path, O_CREAT|O_TRUNC|O_EXCL|O_APPEND|O_WRONLY, 0600)) >= 0) {
					while ((n = read(repFd, buf, sizeof(buf))) > 0) {
						write(fd, buf, n);
					}
					close(fd);
					rename(path, pas->pw_name);
				} else {
					fprintf(stderr, "unable to create %s/%s: %s\n",
							CDir,
							path,
							strerror(errno)
						   );
				}
				close(repFd);
			}
			break;
		case DELETE:
			remove(pas->pw_name);
			break;
		case NONE:
		default:
			break;
	}

	/*
	 *  Bump notification file.  Handle window where crond picks file up
	 *  before we can write our entry out.
	 */

	if (option == REPLACE || option == DELETE) {
		FILE *fo;
		struct stat st;

		while ((fo = fopen(CRONUPDATE, "a"))) {
			fprintf(fo, "%s\n", pas->pw_name);
			fflush(fo);
			if (fstat(fileno(fo), &st) != 0 || st.st_nlink != 0) {
				fclose(fo);
				break;
			}
			fclose(fo);
			/* loop */
		}
		if (fo == NULL) {
			fprintf(stderr, "unable to append to %s/%s\n", CDir, CRONUPDATE);
		}
	}
	exit(0);
	/* not reached */
}

void
printlogf(int level, const char *ctl, ...)
{
	va_list va;
	char buf[LOG_BUFFER];

	va_start(va, ctl);
	vsnprintf(buf, sizeof(buf), ctl, va);
	write(2, buf, strlen(buf));
	va_end(va);
}

void
Usage(void)
{
	/*
	 * parse error
	 */
	printf("crontab " VERSION "\n");
	printf("crontab file [-u user]  replace crontab from file\n");
	printf("crontab -  [-u user]    replace crontab from stdin\n");
	printf("crontab -l [-u user]    list crontab\n");
	printf("crontab -e [-u user]    edit crontab\n");
	printf("crontab -d [-u user]    delete crontab\n");
	printf("crontab -c dir <opts>   specify crontab directory\n");
	exit(2);
}

int
GetReplaceStream(const char *user, const char *file)
{
	int filedes[2];
	int pid;
	int fd;
	int n;
	char buf[RW_BUFFER];

	if (pipe(filedes) < 0) {
		perror("pipe");
		return(-1);
	}
	if ((pid = fork()) < 0) {
		perror("fork");
		return(-1);
	}
	if (pid > 0) {
		/*
		 * PARENT
		 * Read from pipe[0], return it (or -1 if it's empty)
		 */

		close(filedes[1]);
		if (read(filedes[0], buf, 1) != 1) {
			close(filedes[0]);
			filedes[0] = -1;
		}
		return(filedes[0]);
	}

	/*
	 * CHILD
	 * Read from fd for "$file", write to pipe[1]
	 */

	close(filedes[0]);

	if (ChangeUser(user, NULL) < 0)
		exit(0);

	fd = open(file, O_RDONLY);
	if (fd < 0) {
		printlogf(0, "unable to open %s: %s", file, strerror(errno));
		exit(1);
	}
	buf[0] = 0;
	write(filedes[1], buf, 1);
	while ((n = read(fd, buf, sizeof(buf))) > 0) {
		write(filedes[1], buf, n);
	}
	exit(0);
}

void
EditFile(const char *user, const char *file)
{
	int pid;

	if ((pid = fork()) == 0) {
		/*
		 * CHILD - change user and run editor on "$file"
		 */
		const char *ptr;
		char visual[SMALL_BUFFER];

		if (ChangeUser(user, TMPDIR) < 0)
			exit(0);
		if ((ptr = getenv("EDITOR")) == NULL || strlen(ptr) >= sizeof(visual))
			if ((ptr = getenv("VISUAL")) == NULL || strlen(ptr) >= sizeof(visual))
				ptr = PATH_VI;

		/* [v]snprintf write at most size including \0; they'll null-terminate, even when they truncate */
		/* return value >= size means result was truncated */
		if (snprintf(visual, sizeof(visual), "%s %s", ptr, file) < sizeof(visual))
			execl("/bin/sh", "/bin/sh", "-c", visual, NULL);
		printlogf(0, "couldn't exec %s", visual);
		exit(1);
	}
	if (pid < 0) {
		/*
		 * PARENT - failure
		 */
		perror("fork");
		exit(1);
	}
	waitpid(pid, NULL, 0);
}

